### PR TITLE
Prevent closing the app while it's updating

### LINKED ISSUE #5197: update electron-winstaller to help app clean up after itself
## Description
The Squirrel dependency is out of date by a few versions also, it should automatically update itself alongside Git Desktop.

https://github.com/Squirrel/Squirrel.Windows/blob/master/docs/getting-started/1-integrating.md#basic-squirrelwindows-update-code

## Version
1.2.6 on windows 8

### LINKED ISSUE #7055: after closing desktop during update, desktop cannot be reopened
Suddenly my install of Desktop wouldn't open.  It had only been a couple days since I last had it open for just a quick moment.

Checking things, there was something wrong with the .EXE file, but I was able to use one of the backup versions (1.6.3 beta 3) to reupdate.

Looking in my squirrelSetup.log, on 3-08-2019 the log abruptly ended.

```
2019-03-08 23:19:10> DeltaPackageBuilder: lib\net45\resources\app\static\gitignore\Global\Tags.gitignore.diff exists unchanged, skipping
2019-03-08 23:19:10> DeltaPac
2019-03-10 21:12:05> Program: Starting Squirrel Updater: --checkForUpdate https://central.github.com/api/deployments/desktop/desktop/latest?version=1.6.3-beta3&env=beta
2019-03-10 21:12:05> Program: Fetching update information, downloading from https://central.github.com/api/deployments/desktop/desktop/latest?version=1.6.3-beta3&env=beta
2019-03-10 21:12:05> CheckForUpdateImpl: Using existing staging user ID: 9d06e06a-0140-5f85-877a-4997d48d69fd
```

I am guessing the updater was terminated by me closing Desktop.  I have observed this on other machines that were quite old and had a long update list to get Desktop up to date, closing down stopped the 'updater' task and then reopening Desktop would restart the whole update process all over again (from the beginning).

I don't think the update process should be killed upon closing of the Desktop application.

Looking back at the logs it appears this had gone on the previous couple times I had opened Desktop.  Looks like I had 1.6.3 beta 3 on 03-06-2019 and it was first trying to apply beta 4, and also had failed then, probably because it was closed.

OS: Windows 10 (19H1 18353)

### PR DESCRIPTION
**Pull request by sergiou87**
Closes #7055
Closes #5197

## Description
After [digging](https://github.com/Squirrel/Squirrel.Windows/issues/1195#issuecomment-1216633178) into some complaints from users about the app leaving temporary files after updates, and with [a hint](https://github.com/desktop/desktop/issues/11745#issuecomment-1079310876) from one of the said users, I noticed how closing GitHub Desktop while it's being updated can cause either a corrupted installation (unable to open the next time) or just cluttering up the user's harddrive.

This PR attempts to prevent users from closing Desktop while an update is in progress by showing a dialog when the user tries to quit the app. From that dialog, the user might choose to quit the app anyway. Otherwise, the app will quit automatically when the update is completed.

### Screenshots

https://user-images.githubusercontent.com/1083228/194107135-07990e3d-1e3a-4f47-9f37-22df3a3dc878.mov

## Release notes

Notes: [Fixed] Prevent closing the GitHub Desktop while it's being updated

### CODE REVIEW

File: app/src/ui/lib/update-store.ts [hunk 1]
```diff
@@ -56,7 +56,7 @@ export interface IUpdateState {
 /** A store which contains the current state of the auto updater. */
 class UpdateStore {
   private emitter = new Emitter()
-  private status = UpdateStatus.UpdateNotChecked
+  private status = UpdateStatus.UpdateAvailable // UpdateStatus.UpdateNotChecked
```
Comment (niik) [inline app/src/ui/lib/update-store.ts]: lol, let's make triply sure we don't merge this :smile:

File: app/src/main-process/app-window.ts [hunk 2]
```diff
@@ -95,7 +97,40 @@ export class AppWindow {
...
+    })
+
     this.window.on('close', e => {
+      // On macOS, closing the window doesn't mean the app is quitting. If the
+      // app is updating, we will prevent the window from closing only when the
+      // app is also quitting.
+      if (
+        (!__DARWIN__ || quitting) &&
+        !quittingEvenDownloadingUpdate &&
+        // TODO: DON'T MERGE THIS 1!== NaN!!!!
```
Comment (niik) [inline app/src/main-process/app-window.ts]: loool

File: app/src/lib/ipc-shared.ts [hunk 3]
```diff
@@ -46,6 +46,8 @@ export type RequestChannels = {
   'menu-event': (name: MenuEvent) => void
   log: (level: LogLevel, message: string) => void
   'will-quit': () => void
+  'will-quit-even-updating': () => void
```
Comment (niik) [inline app/src/lib/ipc-shared.ts]: Is this missing an `If` or something? I'm having a hard time parsing it `will-quit-even-if-updating` or `will-quite-when-updating` perhaps?

**Review by niik — CHANGES_REQUESTED** [review]
First look at this seems like a decent approach. It does feel a bit weird that a dialog is responsible for quitting the app but I think I can learn to live with it 😄

File: app/src/lib/ipc-shared.ts (hunk 3 above)
Comment (sergiou87) [reply app/src/lib/ipc-shared.ts]: Probably, sorry, my English sucks 😂

**Comment by sergiou87** [comment]
> It does feel a bit weird that a dialog is responsible for quitting the app but I think I can learn to live with it 😄

@niik I intend to go back to the dispatcher -> app store for that, which should be a bit cleaner and more in line with the rest of our dialogs. This is just kind of a PoC, so as long as you don't see any holes in this logic (knowing closing/quitting are different actions on macOS), I'm happy 😂  Thanks for the quick review!! 🙌

**Comment by sergiou87** [comment]
Changed it to look like this:

https://user-images.githubusercontent.com/1083228/194862690-4261bee1-1068-48e8-a2f4-a77b7f2290dd.mov

**Comment by sergiou87** [comment]
After a few more changes, it finally looks like this:
image

File: app/src/main-process/app-window.ts (hunk 2 above)
Comment (sergiou87) [reply app/src/main-process/app-window.ts]: Removed in [c334c20](https://github.com/desktop/desktop/pull/15416/commits/c334c203c83fa116fe7c200b70cc45a480053c9d)!

File: app/src/ui/lib/update-store.ts (hunk 1 above)
Comment (sergiou87) [reply app/src/ui/lib/update-store.ts]: Removed in [c334c20](https://github.com/desktop/desktop/pull/15416/commits/c334c203c83fa116fe7c200b70cc45a480053c9d)!

**Comment by sergiou87** [comment]
I just added a feature flag, but other than that this should be ready for another review @niik 🙏

File: app/src/ui/installing-update/installing-update.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+/**
+ * A dialog that presents information about the
+ * running application such as name and version.
+ */
+export class InstallingUpdate extends React.Component<IInstallingUpdateProps> {
+  private updateStoreEventHandle: Disposable | null = null
+
+  public constructor(props: IInstallingUpdateProps) {
+    super(props)
+  }
```
Comment (niik) [inline app/src/ui/installing-update/installing-update.tsx]: No need for a constructor when we don't have any state

```suggestion
```

**Review by niik — CHANGES_REQUESTED** [review]
Only a stylistic request. One other thing though, should we perhaps call `cancelQuittingApp` from the unmount callback of the InstallingUpdate dialog if the dialog is being unmounted without an explicit choice either way from the user? Like if the user clicks on an `open in Desktop` link on GitHub.com (i.e. `open x-github-desktop-dev-auth://openRepo//`) while the dialog is open?

**Comment by sergiou87** [comment]
That's a very good point!

**Review by niik — APPROVED** [review]

**Pull request merged**